tasks/README.md

```diff
@@ -0,0 +1,244 @@
+
+**input_method**: What input method to use to pass parameters to the task. Default varies, see [Input Methods](#input-methods).
+
+**parameters**:  The parameters or input the task accepts listed with a [Puppet data type](../language/types_values_variables.md) string and optional description. Top level params names have the same restrictions as Puppet class param names and must match `\A[a-z][a-z0-9_]*\z`.
+
+**output**: A description of the fields in the tasks output if it returns structured data. Names for top level output keys must match the regex `\A[a-z][a-z0-9_]*\z`. This exists not to validate output but to generate documentation and provide formatting hints to the UI. Keys beginning with '_' are reserved for internal metaparams (for instance `_output`).
+
+#### Example
+
+```json
+{
+  "description": "Description of what the task does",
+  "parameters": {
+    "param1" : {
+      "description": "Description of what param1 does with option1 or option2",
+      "type": "Enum[option1, option2]"
+    },
+    "param2": {
+      "description": "Description of what param2 does with an array of non-empty strings",
+      "type": "Array[String[0]]"
+    },
+    "param3": {
+      "description": "Description of optional param3",
+      "type": "Optional[Integer]"
+    }
+  }
+}
+```
+
+A JSON schema of metadata accepted by task runners is included in [task.json](task.json). The Forge also hosts https://forgeapi.puppet.com/schemas/task.json describing requirements for metadata in published modules.
+
+## Task parameters
+
+The parameters property is used to validate the parameters to a task and generate UIs.
+
+If the parameters property is missing or `null` any parameter values with be accepted. Authors should not write tasks without specifying parameters.
+
+If the parameters property is empty no parameters will be accepted.
+
+Any parameter that does not specify a type will accept any value (default type is [Any](../language/types_values_variables.md#any) or [Data](../language/types_values_variables.md#data)).
+
+If a parameter type accepts `null` the task runner will accept either a `null` value or the absence of the property. Task authors must accept missing properties for nullable parameters. Task authors must not differentiate between absent properties and properties with `null` values.
+
+## Task execution
+
+The task file is copied to the target and then executed on the target by the task runner.
+
+No arguments are passed to the task when it is executed.
+
+The task file in the module does not need execute permissions set.
+
+The location of the task file varies based on the transport used by the task runner. Task authors have no control over this path.
+
+The environment the task is executed in is also controlled by the task runner. Task authors should document any requirements they have.
```

Comment:
@adreyer `environment` is an overloaded term, do you have suggestions on how to be more specific? Is this system environment variables or general operating environment (as in what's installed, what you have access to)? Doesn't seem like Puppet environment.

Reply:
I think it's primarily the system environment variables. like what's in $PATH etc.